File: heap/MinHeap.java
package heap;

public class MinHeap {
    private int[] heap;
    private int size;

    MinHeap(int capacity) {
        heap = new int[capacity];
        size = 0;
    }

    private int parent(int i) { return (i - 1) / 2; }
    private int leftChild(int i) { return 2 * i + 1; }
    private int rightChild(int i) { return 2 * i + 2; }

    public void insert(int data) {
        heap[size] = data;
        heapifyUp(size);
        size++;
    }

    private void heapifyUp(int i) {
        while (i > 0) {
            int parentIndex = parent(i);
            if (heap[i] < heap[parentIndex]) {
                int temp = heap[i];
                heap[i] = heap[parentIndex];
                heap[parentIndex] = temp;
                i = parentIndex;
            } else {
                break;
            }
        }
    }

    public int extractMin() {
        if (size == 0) return -1;

        int min = heap[0];
        heap[0] = heap[size - 1];
        size--;
        heapifyDown(0);
        return min;
    }

    private void heapifyDown(int i) {
        while (true) {
            int smallest = i;
            int left = leftChild(i);
            int right = rightChild(i);

            if (left < size && heap[left] < heap[smallest]) {
                smallest = left;
            }

            if (right < size && heap[right] < heap[smallest]) {
                smallest = right;
            }

            if (smallest != i) {
                int temp = heap[i];
                heap[i] = heap[smallest];
                heap[smallest] = temp;
                i = smallest;
            } else {
                break;
            }
        }
    }

    public void display() {
        for (int i = 0; i < size; i++) {
            System.out.print(heap[i] + " ");
        }
        System.out.println();
    }
}

File: Python/heap/min_heap.py
class MinHeap:
    def __init__(self):
        self.heap = []

    def _parent(self, index):
        return (index - 1) // 2

    def _left(self, index):
        return 2 * index + 1

    def _right(self, index):
        return 2 * index + 2

    def _swap(self, i, j):
        self.heap[i], self.heap[j] = self.heap[j], self.heap[i]

    def insert(self, value):
        self.heap.append(value)
        index = len(self.heap) - 1

        while index > 0:
            parent_index = self._parent(index)

            if self.heap[parent_index] > self.heap[index]:
                self._swap(index, parent_index)
                index = parent_index
            else:
                break

    def remove_min(self):
        if not self.heap:
            return None

        if len(self.heap) == 1:
            return self.heap.pop()

        min_value = self.heap[0]

        self.heap[0] = self.heap.pop()

        current = 0

        while True:
            left = self._left(current)
            right = self._right(current)

            if left >= len(self.heap):
                break

            smallest = left

            if right < len(self.heap) and self.heap[right] < self.heap[left]:
                smallest = right

            if self.heap[current] <= self.heap[smallest]:
                break

            self._swap(current, smallest)
            current = smallest

        return min_value